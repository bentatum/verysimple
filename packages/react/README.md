# @verysimple/react

This is the `@verysimple/react` package