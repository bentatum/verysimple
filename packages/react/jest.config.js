/** @type {import('jest').Config} */
module.exports = {
  testEnvironment: "jsdom",
  setupFilesAfterEnv: ["@testing-library/jest-dom/extend-expect"],
  collectCoverage: true,
  coverageThreshold: {
    global: {
      statements: 75,
      branches: 72,
      functions: 75,
      lines: 75,
    },
  },
  coverageReporters: ["text-summary", "json-summary"],
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!**/*.d.ts"],
  coverageDirectory: "coverage",
  moduleNameMapper: {
    "@/test/(.*)": "<rootDir>/test/$1",
    "@/(.*)$": "<rootDir>/src/$1",
  },
  modulePathIgnorePatterns: ["<rootDir>/dist/"],
  prettierPath: null,
};
