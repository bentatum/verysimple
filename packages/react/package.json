{
  "name": "@verysimple/react",
  "version": "8.1.1",
  "description": "A very simple component library",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "author": "Ben Tatum <mail@benjtatum.com>",
  "license": "MIT",
  "repository": "https://github.com/bentatum/verysimple",
  "scripts": {
    "test": "jest",
    "prebuild": "rm -rf dist",
    "build": "tsc --project tsconfig.build.json && tsc-alias"
  },
  "peerDependencies": {
    "react": ">=16",
    "react-dom": ">=16"
  },
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@testing-library/dom": "^8.20.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/inquirer": "latest",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.21",
    "@types/testing-library__jest-dom": "^5.14.5",
    "babel-jest": "^29.0.3",
    "babel-loader": "^8.2.5",
    "classnames": "^2.3.2",
    "framer-motion": "^10.15.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "^3.1.8",
    "tsc-alias": "^1.8.2",
    "typescript": "^4.9.3"
  },
  "volta": {
    "node": "20.10.0",
    "yarn": "1.22.21"
  },
  "dependencies": {
    "@headlessui/react": "^2.1.9",
    "@heroicons/react": "^2.0.13",
    "clsx": "^2.1.0",
    "framer-motion": "^10.15.0"
  }
}
