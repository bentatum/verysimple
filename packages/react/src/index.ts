export * from "./ButtonBase";
export * from "./Card";
export * from "./Dialog";
export * from "./FilledButtonBase";
export * from "./FilledIconButton";
export * from "./GradientBorder";
export * from "./helpers";
export * from "./IconButton";
export * from "./NormalIconButton";
export * from "./OutlinedButtonBase";
export * from "./OutlinedIconButton";
export * from "./Spinner";
export * from "./types";
