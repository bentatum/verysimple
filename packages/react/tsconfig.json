{
  "extends": "@verysimple/tsconfig/react-library.json",
  "compilerOptions": {
    "outDir": "./dist",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["node_modules"]
}
