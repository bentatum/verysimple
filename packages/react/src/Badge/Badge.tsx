import classNames from "classnames";
import { ComponentPropsWithRef, forwardRef, ReactNode } from "react";

export interface BadgeProps extends ComponentPropsWithRef<"div"> {
  badgeContent?: ReactNode;
  badgePosition?:
    | "top"
    | "right"
    | "bottom"
    | "left"
    | "top-left"
    | "top-right"
    | "bottom-right"
    | "bottom-left";
}

export const Badge = forwardRef<HTMLDivElement, BadgeProps>(
  (
    { children, badgeContent, badgePosition = "top-right", className = "" },
    ref
  ) => {
    if (!badgeContent) {
      return <>{children}</>;
    }
    const hasOverrides = className.match(
      /top-|right-|bottom-|left-|translate-/
    );
    return (
      <div ref={ref} className="relative">
        {children}
        {badgeContent && (
          <div
            className={classNames(
              "absolute z-10",
              !hasOverrides && {
                "top-0": badgePosition === "top",
                "right-0": badgePosition === "right",
                "bottom-0": badgePosition === "bottom",
                "left-0": badgePosition === "left",
                "left-1/2": ["top", "bottom"].includes(badgePosition),
                "top-1/2": ["left", "right"].includes(badgePosition),
                "translate-x-1/2": [
                  "right",
                  "top-right",
                  "bottom-right",
                ].includes(badgePosition),
                "-translate-x-1/2": [
                  "bottom",
                  "top",
                  "left",
                  "top-left",
                  "bottom-left",
                ].includes(badgePosition),
                "translate-y-1/2": [
                  "bottom",
                  "bottom-left",
                  "bottom-right",
                ].includes(badgePosition),
                "-translate-y-1/2": [
                  "top",
                  "top-left",
                  "top-right",
                  "right",
                  "left",
                ].includes(badgePosition),
                "top-0 right-0": badgePosition === "top-right",
                "top-0 left-0": badgePosition === "top-left",
                "bottom-0 right-0": badgePosition === "bottom-right",
                "bottom-0 left-0": badgePosition === "bottom-left",
              },
              className
            )}
          >
            {badgeContent}
          </div>
        )}
      </div>
    );
  }
);

Badge.displayName = "Badge";

export default Badge;
