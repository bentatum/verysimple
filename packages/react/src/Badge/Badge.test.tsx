import { render, screen } from "@testing-library/react";
import { createRef } from "react";
import Badge from "./Badge";

describe("Badge", () => {
  it("should render", () => {
    const { container } = render(
      <Badge badgeContent="99">
        <button>Notifications</button>
      </Badge>
    );
    expect(container).toMatchSnapshot();
  });

  it("should not render when badge content is empty", () => {
    render(
      <Badge badgeContent={false}>
        <button>Notifications</button>
      </Badge>
    );
    expect(screen.queryByText("Notifications")).toBeInTheDocument();
  });

  describe("badge position", () => {
    it("should render top right by default", () => {
      render(
        <Badge badgeContent="99">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass("top-0", "right-0");
    });
    test("top right", () => {
      render(
        <Badge badgeContent="99" badgePosition="top-right">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass("top-0", "right-0");
    });
    test("top left", () => {
      render(
        <Badge badgeContent="99" badgePosition="top-left">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass("top-0", "left-0");
    });
    test("bottom right", () => {
      render(
        <Badge badgeContent="99" badgePosition="bottom-right">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass("bottom-0", "right-0");
    });
    test("bottom left", () => {
      render(
        <Badge badgeContent="99" badgePosition="bottom-left">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass("bottom-0", "left-0");
    });
    test("top", () => {
      render(
        <Badge badgeContent="99" badgePosition="top">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass(
        "top-0",
        "left-1/2",
        "-translate-x-1/2"
      );
    });
    test("right", () => {
      render(
        <Badge badgeContent="99" badgePosition="right">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass(
        "right-0",
        "top-1/2",
        "-translate-y-1/2"
      );
    });
    test("bottom", () => {
      render(
        <Badge badgeContent="99" badgePosition="bottom">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass(
        "bottom-0",
        "left-1/2",
        "-translate-x-1/2"
      );
    });
    test("left", () => {
      render(
        <Badge badgeContent="99" badgePosition="left">
          <button>Notifications</button>
        </Badge>
      );
      expect(screen.getByText("99")).toHaveClass(
        "left-0",
        "top-1/2",
        "-translate-y-1/2"
      );
    });
    it("can be overriden by className", () => {
      render(
        <Badge
          badgeContent="99"
          badgePosition="bottom-left"
          className="top-1 right-1"
        >
          <button>Notifications</button>
        </Badge>
      );
      const element = screen.getByText("99");
      expect(element).toHaveClass("top-1", "right-1");
      expect(element).not.toHaveClass("bottom-0", "left-0");
    });
  });

  it("forwards the ref", () => {
    const ref = createRef<HTMLDivElement>();
    render(
      <Badge ref={ref} badgeContent="99">
        <button>Notifications</button>
      </Badge>
    );
    expect(ref.current).toBeInTheDocument();
  });

  it("has a display name", () => {
    expect(Badge.displayName).toBe("Badge");
  });
});
