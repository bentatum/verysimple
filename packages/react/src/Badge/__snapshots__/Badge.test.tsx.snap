// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render 1`] = `
<div>
  <div
    class="relative"
  >
    <button>
      Notifications
    </button>
    <div
      class="absolute z-10 translate-x-1/2 -translate-y-1/2 top-0 right-0"
    >
      99
    </div>
  </div>
</div>
`;
